librad/src/net/protocol.rs

```diff
@@ -503,18 +518,32 @@ where
         let remote_id = conn.remote_peer_id();
         tracing::info!(remote.id = %remote_id, "New incoming connection");
 
-        {
-            self.connections.lock().await.insert(remote_id, conn);
+        let mut connections = self.connections.lock().await;
+        if connections.contains_key(&remote_id) {
+            tracing::warn!(remote.id = %remote_id, "rejecting duplicate incoming connection");
+
+            conn.close(CloseReason::DuplicateConnection);
+            drop(incoming);
+
             self.subscribers
-                .emit(ProtocolEvent::Connected(remote_id))
+                .emit(ProtocolEvent::Disconnecting(remote_id))
                 .await;
-        }
+            drop(connections);
```

Comment:
![drop](https://user-images.githubusercontent.com/1585/99960335-dd83b600-2d8c-11eb-8b6e-64df3781adfa.gif)

Reply:
What does this drop do? Is it to guarantee that the mutexguard is dropped?

Reply:
Yes